src/core/fetchers/segment/promise_prioritizer.ts

```diff
@@ -0,0 +1,318 @@
+import log from "../../../log";
+import arrayFindIndex from "../../../utils/array_find_index";
+import PPromise from "../../../utils/promise";
+import TaskCanceller, {
+  CancellationError,
+  CancellationSignal,
+} from "../../../utils/task_canceller";
+
+export default class PromisePrioritizer<T> {
+  private _waitingQueue: Array<IPrioritizerTask<T>>;
+  private _pendingTasks: Array<IPrioritizerTask<T>>;
+  private _prioritySteps: IPrioritizerPrioritySteps;
+
+  constructor({ prioritySteps }: IPrioritizerOptions) {
+    this._waitingQueue = [];
+    this._pendingTasks = [];
+    this._prioritySteps = prioritySteps;
+
+    if (this._prioritySteps.high >= this._prioritySteps.low) {
+      throw new Error(
+        "FP Error: the max high level priority should be given a lower" +
+          "priority number than the min low priority."
+      );
+    }
+  }
+
+  private get _minPendingPriority(): number | null {
+    if (this._pendingTasks.length === 0) {
+      return null;
+    }
+    return Math.min(...this._pendingTasks.map((t) => t.priority));
+  }
+
+  public create(
+    promise: () => Promise<T>,
+    priority: number,
+  ): Promise<T> {
+    let newTask: IPrioritizerTask<T>;
+    return new PPromise<T>((resolve, reject) => {
+      const taskCanceller = new TaskCanceller()
+      const unregisterCancelSignal = taskCanceller.signal.register(
+        (cancellationError: CancellationError) => {
+          // cancel task if running, remove from all queues etc.
+          newTask.trigger(false)
+          this._cleanUpTask(newTask);
+          reject(cancellationError);
```

Comment:
If the internal Promise is interrupted, you reject the whole outer Promise?
I'm not sure that this is what we want